managed/services/management/service.go

```diff
@@ -162,3 +173,111 @@ func (s *ServiceService) validateRequest(request *managementpb.RemoveServiceRequ
+//
+//nolint:unparam
+func (s *ServiceService) ListServices(ctx context.Context, req *managementpb.ListServiceRequest) (*managementpb.ListServiceResponse, error) {
+	filters := models.ServiceFilters{
+		NodeID:        req.NodeId,
+		ServiceType:   convertServiceType(req.ServiceType),
+		ExternalGroup: req.ExternalGroup,
+	}
+
+	var services []*models.Service
+	var agents []*models.Agent
+	var nodes []*models.Node
+
+	agentToAPI := func(agent *models.Agent) *managementpb.UniversalAgent {
+		return &managementpb.UniversalAgent{
+			AgentId:     agent.AgentID,
+			AgentType:   string(agent.AgentType),
+			Status:      agent.Status,
+			IsConnected: s.r.IsConnected(agent.AgentID),
+		}
+	}
+
+	// TODO: provide a higher level of data consistency guarantee by using a locking mechanism.
+	errTX := s.db.InTransaction(func(tx *reform.TX) error {
+		var err error
+		services, err = models.FindServices(tx.Querier, filters)
+		if err != nil {
+			return err
+		}
+
+		agents, err = models.FindAgents(tx.Querier, models.AgentFilters{})
+		if err != nil {
+			return err
+		}
+
+		nodes, err = models.FindNodes(tx.Querier, models.NodeFilters{})
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+
+	if errTX != nil {
+		return nil, errTX
+	}
+
+	nodeMap := make(map[string]string, len(nodes))
+	for _, node := range nodes {
+		nodeMap[node.NodeID] = node.NodeName
+	}
+
+	resultSvc := make([]*managementpb.UniversalService, 0, len(services))
+	for _, service := range services {
+		labels, err := service.GetCustomLabels()
+		if err != nil {
+			return nil, err
+		}
+
+		svc := &managementpb.UniversalService{
+			Address:        pointer.GetString(service.Address),
+			Agents:         []*managementpb.UniversalAgent{},
+			Cluster:        service.Cluster,
+			CreatedAt:      timestamppb.New(service.CreatedAt),
+			CustomLabels:   labels,
+			DatabaseName:   service.DatabaseName,
+			Environment:    service.Environment,
+			ExternalGroup:  service.ExternalGroup,
+			NodeId:         service.NodeID,
+			Port:           uint32(pointer.GetUint16(service.Port)),
+			ReplicationSet: service.ReplicationSet,
+			ServiceId:      service.ServiceID,
+			ServiceType:    string(service.ServiceType),
+			ServiceName:    service.ServiceName,
+			Socket:         pointer.GetString(service.Socket),
+			UpdatedAt:      timestamppb.New(service.UpdatedAt),
+		}
+
+		nodeName, ok := nodeMap[service.NodeID]
+		if ok {
+			svc.NodeName = nodeName
+		}
+
+		var svcAgents []*managementpb.UniversalAgent
+
+		for _, agent := range agents {
+			if IsNonExporterAgent(agent, service) {
+				svcAgents = append(svcAgents, agentToAPI(agent))
+			}
+
+			if IsVMAgent(agent, service) {
+				svcAgents = append(svcAgents, agentToAPI(agent))
+			}
+
+			if IsExporterAgent(agent, service) {
+				svcAgents = append(svcAgents, agentToAPI(agent))
+			}
+
+			svc.Agents = svcAgents
+		}
+
+		resultSvc = append(resultSvc, svc)
```

Comment:
```suggestion
	resultSvc := make([]*managementpb.UniversalService, len(services))
	for i, service := range services {
		labels, err := service.GetCustomLabels()
		if err != nil {
			return nil, err
		}

		svc := &managementpb.UniversalService{
			Address:        pointer.GetString(service.Address),
			Agents:         []*managementpb.UniversalAgent{},
			Cluster:        service.Cluster,
			CreatedAt:      timestamppb.New(service.CreatedAt),
			CustomLabels:   labels,
			DatabaseName:   service.DatabaseName,
			Environment:    service.Environment,
			ExternalGroup:  service.ExternalGroup,
			NodeId:         service.NodeID,
			Port:           uint32(pointer.GetUint16(service.Port)),
			ReplicationSet: service.ReplicationSet,
			ServiceId:      service.ServiceID,
			ServiceType:    string(service.ServiceType),
			ServiceName:    service.ServiceName,
			Socket:         pointer.GetString(service.Socket),
			UpdatedAt:      timestamppb.New(service.UpdatedAt),
		}

		nodeName, ok := nodeMap[service.NodeID]
		if ok {
			svc.NodeName = nodeName
		}

		var svcAgents []*managementpb.UniversalAgent

		for _, agent := range agents {
			if IsNonExporterAgent(agent, service) {
				svcAgents = append(svcAgents, agentToAPI(agent))
			}

			if IsVMAgent(agent, service) {
				svcAgents = append(svcAgents, agentToAPI(agent))
			}

			if IsExporterAgent(agent, service) {
				svcAgents = append(svcAgents, agentToAPI(agent))
			}

			svc.Agents = svcAgents
		}

		resultSvc[i] = svc
```

Reply:
Not a big difference but shows that we already know the exact size of the resulting array.